### ISSUE TITLE
More scalable memory management

### ISSUE THREAD

**Issue by wlandau** [ropensci/drake#897]
## Prework

- [x] Read and abide by `drake`'s [code of conduct](https://github.com/ropensci/drake/blob/master/CODE_OF_CONDUCT.md).
- [x] Search for duplicates among the [existing issues](https://github.com/ropensci/drake/issues), both open and closed.

## Motivation

`drake` automatically loads all a target's dependencies into memory. This does not bode well for all use cases, such as large `combine()` steps.

``` r
library(drake)

plan <- drake_plan(
    large_output = target(
        run_mcmc(iterations = 1e6, rep = r),
        transform = map(r = !!seq_len(1e2))
    ),
    huge_output = target(
        dplyr::bind_rows(large_output),
        transform = combine(large_output)
    )
)

config <- drake_config(plan)
vis_drake_graph(config)
```

![](https://i.imgur.com/Dd5lXEZ.png)

<sup>Created on 2019-06-05 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

`drake` loads all the `large_output` data into memory to compute `huge_output`. Not a good idea.

## Proposal

- [x] `make(memory_strategy = "none")` should not try to load dependencies.
- [x] A new custom `drake_plan()` column should allow users to set per-target memory strategies, e.g. `drake_plan(x = target(f(), memory_strategy = "none"), y = target(g(), memory_strategy = "lookahead"))`.
- [ ] A recommendation of how to use the new memory strategy effectively in combine(). Sketch:

```r
plan <- drake_plan(
	large_output = target(
		run_mcmc(iterations = 1e6, rep = r),
		transform = map(r = !!seq_len(1e2))
	),
	huge_output = target(
		combine_datasets(inputs = c(large_output), output_file = file_out("output.csv")),
		transform = combine(large_output)
	)
)
```

where

```r
combine_datasets <- function(inputs, output_file) {
  # Use drake's cache manually.
  cache <- drake_cache()

  # Get the names of the inputs as character strings.
  inputs <- match.call()$inputs %>%
    as.character()

  # Load the inputs one by one and clear them from memory when done.
  for (input in inputs) {
    dataset <- cache$get(
      key = input,
      use_cache = FALSE # Important for memory management.
    )

    # Append manually to the output file.
    write(dataset, output_file, append = TRUE)
  }
}
```

**Comment by wlandau** [comment]
You know what? It may just be easier to use drake’s existing lazy loading capabilities. Details and profiling to follow.

**Comment by wlandau** [comment]
No, lazy loading won't cover this easily enough. We should stick to https://github.com/ropensci/drake/issues/897#issue-452594151 and recommend the use of `readd()`. Probably should talk about memory management in the [manual](https://github.com/ropenscilabs/drake-manual) too.

**Comment by wlandau** [comment]
Mostly implemented in [a branch](https://github.com/ropensci/drake/tree/897) now. Just need proper testing.

- [ ] Repurpose [this test](https://github.com/ropensci/drake/blob/69183c26319872a6cfeeafa44e961e3f9356658a/tests/testthat/test-memory.R#L18-L107) for each memory strategy individually.
- [ ] Test the "unload" and "none" memory strategies in their own test.
    - [ ] `make()` should throw an error if a dependency is not loaded manually.
    - [ ] This error should be fixable if the user invokes `readd()` in the command manually.

**Comment by wlandau** [comment]
For the "unload" and "none" memory strategies, we should also avoid assigning newly-built targets to the evaluation environment where targets are built (`config$eval`). We can do this in `assign_to_envir()`.

**Comment by wlandau** [comment]
Also need to check individual target-level memory management strategies.

**Comment by wlandau** [comment]
Also document in the help file:
- [ ] The new memory strategies.
- [ ] When newly-built targets are retained in the evaluation environment.

**Comment by wlandau** [comment]
Implemented in #900. Need to update the manual.
